### ISSUE TITLE
[Feature Request] Cast / DLNA support?

### ISSUE THREAD

**Issue by sunjaxx** [UnicornsOnLSD/finamp#50]
Any plan to support cast (chromecast or DLNA), something like the native "play on" feature of the official jellyfin app? Thanks!

**Comment by UnicornsOnLSD** [comment]
A lot of people have asked for this, I'm waiting on the audio package I'm using to support it. It's in progress there, so I should be able to add casting soon: https://github.com/ryanheise/just_audio/issues/211

As for DLNA, I might be able to make it work. I don't know much about DLNA, but I've heard that the Jellyfin server should handle it and I just have to give the command to the server.

**Comment by UnicornsOnLSD** [comment]
On iOS, you can use AirPlay to play both downloaded and streamed songs by selecting the device in control center/the playing notification. Still no Cast/DLNA yet though.

**Comment by seniorm0ment** [comment]
I use the Mopidy-Jellyfin plugin for Mopidy, and this would be nice when the audio package is finished :)

**Comment by misterashley** [comment]
> Any plan to support cast (chromecast or DLNA), something like the native "play on" feature of the official jellyfin app? Thanks!

Any updates here? Is there anything someone can do to assist? This would be a killer feature for me, as I try to de-Sp**ify my life.

**Comment by seniorm0ment** [comment]
I know this is reliant on the package, but any updates on this?
Sucks to need to go to the Jellyfin app just to cast to my Mopidy server.

**Comment by sunjaxx** [comment]
Any update/timelines for this? :) Thanks

**Comment by seniorm0ment** [comment]
A temporary solution could be to allow controlling of other Jellyfin devices. While not exactly the same as casting, I don't see why this would not be possible to implement.
It is already a feature with other Jellyfin clients.

**Comment by provokateurin** [comment]
Hi, I'm very interested in this feature and would be willing to develop it. I have no experience with DLNA, Jellyfin APIs or Finamp development, but I think it should be doable. @jmshrv https://api.jellyfin.org/#tag/Dlna and https://api.jellyfin.org/#tag/DlnaServer would be the relevant APIs, right?

**Comment by provokateurin** [comment]
But the casting thing is different from DLNA, right? Because casting would be from the device and DLNA from the server where jellyfin runs if I understood things correctly.

**Comment by provokateurin** [comment]
Ah I see `audio_cast` supports DLNA itself, so it goes from the device? I think both modes of DLNA make sense to be implemented. My usecase would favour the DLNA from the server, so I'd like to implement that part.

**Comment by jmshrv** [comment]
Ah yeah I just mentioned Chromecast as a separate thing, I'm pretty sure DLNA should be handled from the server

**Comment by provokateurin** [comment]
Ok, should we split it up into a separate issue? I think the UI parts will probably be the same, but the logic is obviously completely different.

**Comment by jmshrv** [comment]
Yeah that's a good idea. For implementing DLNA, these are probably the relevant files:

* `services/jellyfin_api.dart` - This actually handles sending HTTP requests to and from Jellyfin. It's pretty much just a simple Chopper thing that returns JSON.
* `services/jellyfin_api_helper.dart` - This is what the whole app uses to interact with Jellyfin (and by extension the offline stuff, but that's all obscured). In this, you'll call the method you make in jellyfin_api.dart and return a nice class instead of some random JSON
* `services/music_player_background_task.dart` - This is where the music service for Finamp is located - it handles current play state. It's closely related to the `audio_service` package, which tells the OS about the currently playing song (for background playback handling, now playing status etc). I don't really know how DLNA will tie in with it for a good user experience, but off the top of my head making a new music service that uses DLNA instead of actually playing audio would probably work.

**Comment by jmshrv** [comment]
Oh, also `models/jellyfin_models.dart`, which contains all the classes that Jellyfin returns. For DLNA, you shouldn't have to mess about with Hive (used for offline storage), so it should be pretty recognisable if you've ever worked with `json_annotation`.

**Comment by provokateurin** [comment]
Thanks for the hints :)

**Comment by provokateurin** [comment]
New issue for DLNA opened, please rename this one to Cast only :)

**Comment by vladiskuz** [comment]
Does anybody start working on that?

**Comment by eugenialoli** [comment]
Any news on Chromecast support? The only reason why I use jellyfin is just so I can play my songs to our hi-fi system in our living room, that has Cast support. I never listen music through my phone, I prefer to use Finamp as a control app for casting. Unfortunately, the official Jellyfin app is extremely buggy when it comes to casting, and it makes it totally unusable for me. I wonder if Finamp has any plans for some solid casting.

**Comment by Chaphasilor** [comment]
Next step towards casting is implementing the second half of Jellyfin's "Play On" feature, which then allows Finamp to control other clients *via the Jellyfin server*. That should add support for "casting" to headless clients and potentially DLNA clients (not sure how exactly this is handled by the server).

If that works well, we should be able to add Chromecast support.

@DuckManQuack how exactly does SqueezeLite receive media? Does it support regular Chromecast?

**Comment by DuckManQuack** [comment]
From what I know its through the old "Logitech Media Server" or Lyrion Music Server now. As for how it gets to LMS I'm not too sure.

**Comment by thornygravy** [comment]
Definitely would be a great addition to such an already feature rich and amazing app!

**Comment by actarus-magnus** [comment]
+1 for this, with a specific use case that might help frame the priority.

The most valuable form of this for a music client isn't really Chromecast/DLNA, but Finamp acting as a *controller* of another Jellyfin session — the "Play On" sender role, essentially Spotify Connect. The target would be a headless server-side player such as jellyfin-mpv-shim connected to a USB DAC.

Why this is compelling:

- Finamp already implements the *controllable* side (it can be controlled by the official apps), so a lot of the session/websocket plumbing already exists — this is the complementary direction.
- It's bit-perfect by design: Finamp would only send commands and item IDs, while the server-side player direct-plays the original FLAC to the DAC. No transcoding, and none of the quality cap you get with AirPlay.
- It delivers the "one beautiful app" experience: local playback on the phone when you're out, then hand the queue over to the home system when you get home — exactly Spotify Connect, but for self-hosted music.

My setup is iPhone + a Jellyfin server running mpv-shim into an external DAC. Today I have to switch to the official Jellyfin app just to cast, which is a poor experience for music. A "Connect"-style device picker inside Finamp would let me stay in one app the whole time.

Thanks for all the work on Finamp. I'm not a developer, but I'd be glad to help test this on a real mpv-shim + DAC setup if anyone picks it up.

**Comment by juxuanu** [comment]
I totally agree with @actarus-magnus . The DIY Chromecast for Jellyfin is having a headless server with mpv-shim (like a Raspberry Pi with the DAC HAT) and selecting the device in the official Jellyfin app.
Being able to set all that up with Finamp would be awesome.

**Comment by actarus-magnus** [comment]
I've implemented this — PR #1642 if anyone wants to take a look.
Tested on a real jellyfin-mpv-shim + USB DAC setup (bit-perfect FLAC 24/96). The handoff preserves playback position in both directions (phone → DAC and back), transport controls and seek work remotely, and there's an automatic fallback to local if the remote session vanishes.
One known gap left as a follow-up: lock-screen/system media controls still drive local playback while remote.

**Comment by Danie10** [comment]
@actarus-magnus thanks yes working for me to Google home speakers across LAN.

**Comment by camblering** [comment]
Oh wow, does this mean Chromecast support is imminent? I’ve been hanging out for this on iOS for a Jellyfish client and I love Finamp.

**Comment by Chaphasilor** [comment]
@camblering Chromecast would be the next step. What @actarus-magnus did now is complete support for Jellyfin's "Play On" mechanism, where clients can talk to the server to control the playback of other clients. This of course then only works with Jellyfin clients.  
But the code infrastructure needed for this (e.g. Finamp showing playback info and a queue that is not played on the local device, but on a remote device), would also be needed for Chromecast support.

Timeline wise I can't give you any thing concrete yet. The PR mentioned above is purely vibe-coded at this point, and while it's seems to be working mostly fine, we will need to make a thorough review pass over the code, and possibly refactor parts of this to ensure it's maintainable, extensible (e.g. Chromecast), and covers all edge cases. I don't know how long this will take, especially since any such refactoring work will most likely be done manually.  

There's also @ember-ana working on Chromecast support specifically, independent form the other PR:  
https://github.com/ember-ana/finamp/commits/re/google-cast-support/  
There's no open PR for that yet, and I'm not sure what the current state of that implementation is, but we will probably end up integrating both these contributions into Finamp somehow, to get full Play On and Google Chromecast support.

**Comment by ember-ana** [comment]
Hiya, I'm the one working on the Google Cast support :)

> I'm not sure what the current state of that implementation is

So, the Google Cast side of things should be all done; as in, there's a `GoogleCast` singleton (and its associated data types etc.) in `lib/services/google_cast.dart` that should be able to handle everything (playback, remote queue management, volume control, etc.) as long as it gets given the data from Finamp.

What's left to do is wiring all that up to Finamp's UI and internal queue/audio handling.

The underlying storage for my home Jellyfin instance had some data loss, and I've been very busy irl, so I haven't found the motivation to set all of that back up. Which means I probably won't be working on things for a while. If someone wants to merge in my progress so far (which should affect the rest of Finamp very little), or pick up where I left off, feel free to :)

**Comment by jedinjapan** [comment]
For those with home audio streaming setups, the Spotify Connect protocol, as well as Squeezelite from LMS are both common targets for playback devices that it would be fantastic to see Finamp support.

I would love to let go of Plexamp in favor of this app, but I'm able to cast to Plex Connect devices in Music Assistant, which makes it preferable for now.
